src/llmcompressor/modeling/gpt_oss.py

```diff
@@ -0,0 +1,249 @@
+    ) -> torch.Tensor:
+        """
+        Implements the MoE computation using the router outputs.
+
+        This is compatible with the GPT-OSS MoE call pattern:
+            experts(hidden_states, router_indices, routing_weights)
+        """
+        assert routing_weights is not None and router_indices is not None, "router inputs required"
+
+        # Normalize shapes to [tokens, H], [tokens, top_k], [tokens, E]
+        if hidden_states.dim() == 3:
+            B, T, H = hidden_states.shape
+            x = hidden_states.reshape(-1, H)
+        else:
+            # Already flattened
+            B, T = 1, hidden_states.shape[0]
+            H = hidden_states.shape[-1]
+            x = hidden_states
+
+        if router_indices.dim() == 3:
+            router_indices = router_indices.reshape(-1, router_indices.shape[-1])
+        if routing_weights.dim() == 3:
+            routing_weights = routing_weights.reshape(-1, routing_weights.shape[-1])
+
+        num_experts_plus_dummy = routing_weights.shape[1]
+        out = torch.zeros_like(x)
+
+        # GPT-OSS router uses an extra "no expert" bucket at index E
+        with torch.no_grad():
+            expert_mask = torch.nn.functional.one_hot(
+                router_indices, num_classes=num_experts_plus_dummy
+            ).permute(2, 1, 0)
+            expert_hit = torch.greater(expert_mask.sum(dim=(-1, -2)), 0).nonzero()
+
+        for idx in expert_hit:
+            e = idx[0].item()
+            if e == self.num_experts:
+                # Skip "no expert" bucket
+                continue
+
+            _, token_idx = torch.where(expert_mask[e])
+            xi = x[token_idx]
+
+            expert = self.experts[e]
+            yi = expert(xi)
+
+            w = routing_weights[token_idx, e, None]
+            out.index_add_(0, token_idx, (yi * w).to(out.dtype))
+
+        return out.view(B, -1, H)
+
+
+@dataclass
+class ExpertMeta:
+    path: str
+    hidden_size: int
+    intermediate_size: int
+    num_experts: int
+    device: torch.device
+    dtype: torch.dtype
+
+
+def get_module_by_path(root: nn.Module, dotpath: str) -> nn.Module:
+    m: nn.Module = root
+    if not dotpath:
+        return root
+    for p in dotpath.split("."):
+        m = getattr(m, p)
+    return m
+
+
+def set_module_by_path(root: nn.Module, dotpath: str, new_module: nn.Module) -> None:
+    parts = dotpath.split(".")
+    parent = get_module_by_path(root, ".".join(parts[:-1]))
+    setattr(parent, parts[-1], new_module)
+
+
+def find_experts(model: nn.Module) -> List[ExpertMeta]:
+    """
+    Locate GPT-OSS MoE expert modules under model.model.layers[*].mlp.experts.
+    """
+    metas: List[ExpertMeta] = []
+    for li, layer in enumerate(model.model.layers):
+        experts = layer.mlp.experts
+        device = next(experts.parameters(), torch.zeros(())).device
+        dtype = next(experts.parameters(), torch.zeros(())).dtype
```

Comment:
I think that we should error if the experts ever don't have parameters, not default to cpu